mdbook-tera-backend/src/tera_renderer/renderer.rs

```diff
@@ -0,0 +1,177 @@
+        }
+        let file_content = std::fs::read_to_string(path)?;
+        let output = self.render_file_content(&file_content, path)?;
+        Ok(std::fs::write(path, output)?)
+    }
+
+    /// Creates the rendering context to be passed to the templates.
+    ///
+    /// # Arguments
+    ///
+    /// `path`: The path to the file that will be added as extra context to the renderer.
+    fn create_context(&mut self, path: &Path) -> Result<tera::Context> {
+        let mut context = tera::Context::new();
+        context.insert("path", path);
+        context.insert("ctx", &serde_json::to_value(&self.ctx)?);
+        context.insert("book_dir", &self.ctx.destination.parent().unwrap());
+
+        Ok(context)
+    }
+
+    /// Rendering logic for an individual file.
+    fn render_file_content(&mut self, file_content: &str, path: &Path) -> Result<String> {
+        let tera_context = self.create_context(path)?;
+
+        let rendered_file = self
+            .tera_template
+            .render_str(file_content, &tera_context)
+            .map_err(|e| anyhow!("Error rendering file {path:?}: {e:?}"))?;
+        Ok(rendered_file)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use tempdir::TempDir;
+
+    use super::*;
+    use crate::tera_renderer::custom_component::TeraRendererConfig;
+
+    const RENDER_CONTEXT_STR: &str = r#"
+    {
+        "version":"0.4.32",
+        "root":"",
+        "book":{
+           "sections": [],
+           "__non_exhaustive": null
+        },
+        "destination": "",
+        "config":{
+           "book":{
+              "authors":[
+                 "Martin Geisler"
+              ],
+              "language":"en",
+              "multilingual":false,
+              "src":"src",
+              "title":"Comprehensive Rust 🦀"
+           },
+           "build":{
+              "build-dir":"book",
+              "use-default-preprocessors":true
+           },
+           "output":{
+              "tera-backend": {
+                    "template_dir": "templates"
+              },
+              "renderers":[
+                 "html",
+                 "tera-backend"
+              ]
+           }
+        }
+     }"#;
+
+    const HTML_FILE: &str = r#"
+        <!DOCTYPE html>
+            {%include "test_template.html" %}
+        </html>
+    "#;
+
+    const TEMPLATE_FILE: &str = "RENDERED";
+
+    const RENDERED_HTML_FILE: &str = r#"
+        <!DOCTYPE html>
+            RENDERED
+        </html>
+    "#;
+
+
+    #[test]
+    fn test_renderer() {
+        let mut ctx = RenderContext::from_json(RENDER_CONTEXT_STR.as_bytes()).unwrap();
+
+        let tmp_dir = TempDir::new("output").expect("Failed to create temp dir");
+        let html_path = tmp_dir.path().join("html");
+        let templates_path = tmp_dir.path().join("templates");
+
+        std::fs::create_dir(&html_path).expect("Failed to create html dir");
+        std::fs::create_dir(&templates_path).expect("Failed to create html dir");
```

Comment:
Please drop the `.expect()` calls (because they're using the wrong strings). Just use `?` and let your test return a `Result`.

Reply:
I disagree on that, the error message is really not useful if we return an anyhow result, even with .context().

Reply:
It's a test -- if it fails, then the developer will have to go and debug it by hand. So we don't need to make this "user friendly" since the user is a developer and since they will have to debug it anyway.

Reply:
But what's the advantage of returning a result here, if it does the same thing with worse error messages?

Reply:
> But what's the advantage of returning a result here, if it does the same thing with worse error messages?

My thinking is that it'll lead to less code in the tests. Test have to be _super clear_ since they're the thing that keep the of the code working :smile: So every line there is a liability — there is nothing to check or validate the line except you and me.

In this case, the code was

```rust
        std::fs::create_dir(&html_path).expect("Failed to create html dir");
        std::fs::create_dir(&templates_path).expect("Failed to create html dir");
```
which is wrong: the error messages are the same but they do different things. So that's a problem and the problem can be avoided (by construction) if you write less code with a `?`. Then you cannot make this mistake at all.

The test will , it will only break when someone is refactoring the code in question. So that person will already be in the middle of a development cycle and so having the test fail is hopefully be met with "Ah, yes of course! I'll have to adjust this now!"

Reply:
The error messages are also off below:

```rust
        std::fs::write(templates_path.join("test_template.html"), TEMPLATE_FILE)
            .expect("Failed to create test.html");
```

and this just shows that it's annoying to keep things in sync :smile: So having less things to keep in sync should be an overall win for everbody.

Reply:
One of our internal [Testing on the Toilet](https://testing.googleblog.com/2007/01/introducing-testing-on-toilet.html) episodes touched upon this: tests should avoid complexity at all cost. They should be obviously correct and don't try to factor common code out into helpers and constants.

So I would even suggest removing the small string constants above and just inline the content below.